flask/logging.py

```diff
@@ -1,94 +1,69 @@
-# -*- coding: utf-8 -*-
-"""
-    flask.logging
-    ~~~~~~~~~~~~~
-
-    Implements the logging support for Flask.
-
-    :copyright: (c) 2015 by Armin Ronacher.
-    :license: BSD, see LICENSE for more details.
-"""
-
 from __future__ import absolute_import
 
+import logging
 import sys
 
 from werkzeug.local import LocalProxy
-from logging import getLogger, StreamHandler, Formatter, getLoggerClass, \
-     DEBUG, ERROR
-from .globals import _request_ctx_stack
-
 
-PROD_LOG_FORMAT = '[%(asctime)s] %(levelname)s in %(module)s: %(message)s'
-DEBUG_LOG_FORMAT = (
-    '-' * 80 + '\n' +
-    '%(levelname)s in %(module)s [%(pathname)s:%(lineno)d]:\n' +
-    '%(message)s\n' +
-    '-' * 80
-)
+from .globals import request
 
 
 @LocalProxy
-def _proxy_stream():
-    """Finds the most appropriate error stream for the application.  If a
-    WSGI request is in flight we log to wsgi.errors, otherwise this resolves
-    to sys.stderr.
+def wsgi_errors_stream():
+    """Find the most appropriate error stream for the application. If a request
+    is active, log to ``wsgi.errors``, otherwise use ``sys.stderr``.
+
+    If you configure your own :class:`logging.StreamHandler`, you may want to
+    use this for the stream. If you are using file or dict configuration and
+    can't import this directly, you can refer to it as
+    ``ext://flask.logging.wsgi_errors_stream``.
     """
-    ctx = _request_ctx_stack.top
-    if ctx is not None:
-        return ctx.request.environ['wsgi.errors']
-    return sys.stderr
+    return request.environ['wsgi.errors'] if request else sys.stderr
+
+
+def has_level_handler(logger):
+    """Check if there is a handler in the logging chain that will handle the
+    given logger's :meth:`effective level <~logging.Logger.getEffectiveLevel>`.
+    """
+    level = logger.getEffectiveLevel()
+    current = logger
+
+    while current:
+        if any(handler.level <= level for handler in current.handlers):
```

Comment:
These effective levels are tricky to think about 😅 . Had to go back and forth with the docs for awhile. `handler.level` is the _lowest_ level that will be handled by the logger, so this seems right.